source/celview.cpp

```diff
@@ -136,73 +138,149 @@ void CelView::framePixelClicked(unsigned x, unsigned y)
-            return; // no new frame -> done
+            this->sendAddFrameCmd(mode, this->currentFrameIndex + nextIdx, imagefilePaths[i]);
+            nextIdx++;
         }
     }
+}
+
+void CelView::insertImageFile(int frameIdx, const QImage img)
+{
+    int prevFrameCount = this->gfx->getFrameCount();
+
+    // insert group with last frame that was deleted in it, or
+    // insert a frame in a group where it was before
+    if (!removedGroupIdxs.empty()) {
+        int removedGroupIdx = removedGroupIdxs.top();
+        this->gfx->insertGroup(removedGroupIdx, frameIdx, img);
+        removedGroupIdxs.pop();
+    } else {
+        int groupIdx = this->removedFrameGroupIdxs.top();
+        this->gfx->insertFrameInGroup(frameIdx, groupIdx, img);
+        this->removedFrameGroupIdxs.pop();
+    }
+
+    int deltaFrameCount = this->gfx->getFrameCount() - prevFrameCount;
+    if (deltaFrameCount == 0) {
+        return; // no new frame -> done
+    }
+    // jump to the first appended frame
+    this->currentFrameIndex = frameIdx;
+    this->updateGroupIndex();
+
     // update the view
     this->initialize(this->gfx);
     this->displayFrame();
 }
 
-void CelView::insertFrame(IMAGE_FILE_MODE mode, int index, const QString &imagefilePath)
+void CelView::insertFrames(int startingIndex, const std::vector<QImage> &images)
 {
-    QImageReader reader = QImageReader(imagefilePath);
-    int numImages = 0;
+    int prevFrameCount = this->gfx->getFrameCount();
 
-    while (true) {
-        QImage image = reader.read();
-        if (image.isNull()) {
-            break;
-        }
-        this->gfx->insertFrame(index + numImages, image);
-        numImages++;
+    for (int idx = 0; idx < images.size(); idx++) {
+        this->gfx->insertFrame(startingIndex + idx, images[idx]);
     }
 
-    if (mode != IMAGE_FILE_MODE::AUTO && numImages == 0) {
-        QMessageBox::critical(this, "Error", "Failed read image file: " + imagefilePath);
+    int deltaFrameCount = this->gfx->getFrameCount() - prevFrameCount;
+    if (deltaFrameCount == 0) {
+        return; // no new frame -> done
     }
+
+    this->currentFrameIndex = startingIndex;
+    this->updateGroupIndex();
+
+    // update the view
+    this->initialize(this->gfx);
+    this->displayFrame();
 }
 
-void CelView::replaceCurrentFrame(const QString &imagefilePath)
+void CelView::removeFrames(int startingIndex, int endingIndex)
 {
-    QImage image = QImage(imagefilePath);
+    int idx = startingIndex;
+    while (startingIndex != endingIndex) {
+        this->removeCurrentFrame(idx);
+        startingIndex++;
+    }
+}
 
-    if (image.isNull()) {
+void CelView::sendAddFrameCmd(IMAGE_FILE_MODE mode, int index, const QString &imagefilePath)
+{
+    AddFrameCommand *command;
+    try {
+        command = new AddFrameCommand(mode, index, imagefilePath);
+    } catch (...) {
+        QMessageBox::critical(this, "Error", "Failed to read image file: " + imagefilePath);
+        delete command;
```

Comment:
If an exception is thrown in the `AddFrameCommand()` constructor, the pointer won't be assigned to `command`. The only possibility here is deleting an uninitialized pointer, which is UB.

```suggestion
```

Reply:
Done, thanks, good to learn new things :D